### PR TITLE
[IMP] spreadsheet: add header grouping feature

### PR DESCRIPTION
**Pull request by hokolomopo**
## Description:

With this commit it is now possible to group columns/rows, and to easily
fold/unfold these groups.

Odoo task ID : [3272931](https://www.odoo.com/web#id=3272931&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

**Review by LucasLefevre — COMMENTED** [review]
Nice work and nice feature :)
I expect some change when moving fold/unfold from UI to core

File: src/actions/view_actions.ts [hunk 1]
```diff
@@ -217,6 +218,64 @@ export const createRemoveFilter: ActionSpec = {
...
+export const groupRows: ActionSpec = {
+  name: (env) => {
+    const selection = env.model.getters.getSelectedZone();
+    if (selection.top === selection.bottom) {
+      return _lt("Group row %s", String(selection.top + 1));
+    }
+    return _lt("Group rows %s - %s", String(selection.top + 1), String(selection.bottom + 1));
+  },
+  execute: (env) => groupHeadersAction(env, "ROW"),
+  isVisible: isSingleZoneSelected,
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/actions/view_actions.ts:247]: Having both group & ungroup is relevant because of all the magic. I can disable the "group" item when there is already a group with this selection tho.

File: src/components/header_group/header_group.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import { SpreadsheetChildEnv } from "../../types/env";
+import { css, cssPropertiesToCss } from "../helpers";
+
+interface Props {
+  group: HeaderGroup;
+  layerOffset: number;
+  openContextMenu(position: DOMCoordinates, menuItems: Action[]): void;
+}
+
+interface GroupBox extends Rect {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/components/header_group/header_group.ts:21]: I do'nt think it should be `GroupRect`, we have too many stuff inside the Object to call it a Rect. We could argue that it shouldn't even extend Rect, but have a groupRect attributes inside of it.

File: src/plugins/ui_stateful/header_grouping_ui.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        }
+      }
+    }
+    return false;
+  }
+
+  isColFolded(sheetId: UID, col: HeaderIndex) {
+    const groups = this.getters.getHeaderGroups(sheetId, "COL");
+    // ADRM DISCUSS: might become a performance issue if there are a lot of groups (this is called by isRowHidden).
+    // Worth maintaning a list of all the hidden rows ?
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/ui_stateful/header_grouping_ui.ts:132]: Doesn't look  like a real issue. Will leave the comment for future reference if it ever becomes a problem.

File: tests/collaborative/ot/ot_rows_added.test.ts [hunk 4]
```diff
@@ -97,6 +101,34 @@ describe("OT with ADD_COLUMNS_ROWS with dimension ROW", () => {
     });
   });
 
+  describe.each(OT_TESTS_ZONE_DEPENDANT_COMMANDS)("zone dependant commands", (cmd) => {
+    test(`add rows after ${cmd.type}`, () => {
+      const command = { ...cmd, sheetId, zone: toZone("A1:C1") };
+      const result = transform(command, addRowsAfter);
+      expect(result).toEqual(command);
+    });
+    test(`add rows after ${cmd.type}`, () => {
```
Comment (LucasLefevre) [inline tests/collaborative/ot/ot_rows_added.test.ts:110]:
```suggestion
    test(`add rows before ${cmd.type}`, () => {
```

File: tests/collaborative/ot/ot_rows_added.test.ts [hunk 5]
```diff
@@ -97,6 +101,34 @@ describe("OT with ADD_COLUMNS_ROWS with dimension ROW", () => {
...
+      const command = { ...cmd, sheetId, zone: toZone("A10:B11") };
+      const result = transform(command, addRowsAfter);
+      expect(result).toEqual({ ...command, zone: toZone("A12:B13") });
+    });
+    test(`add rows after in ${cmd.type}`, () => {
+      const command = { ...cmd, sheetId, zone: toZone("A5:B6") };
+      const result = transform(command, addRowsAfter);
+      expect(result).toEqual({ ...command, zone: toZone("A5:B6") });
+    });
+    test(`add rows before in ${cmd.type}`, () => {
```
Comment (LucasLefevre) [inline tests/collaborative/ot/ot_rows_added.test.ts:120]: those two tests are actually the same. I guess you were trying to test row addition inside a group (and the group should grow)

**Review by LucasLefevre — COMMENTED** [review]
I only partially read the tests for now

File: src/components/grid/grid.ts [hunk 6]
```diff
@@ -652,6 +672,108 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
...
+        });
+    }
+  }
+
+  private processHeaderGroupingEventOnWholeSheet(direction: Direction) {
+    const sheetId = this.env.model.getters.getActiveSheetId();
+    if (direction === "up") {
+      for (const dimension of ["ROW", "COL"] as Dimension[]) {
+        this.env.model.dispatch("COLLAPSE_ALL_HEADER_GROUPS", { sheetId, dimension });
+      }
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/components/grid/grid.ts:730]: That was a big fly :roll_eyes:

File: src/components/header_group/header_group_container.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+  openContextMenu(position: DOMCoordinates, menuItems: Action[]) {
+    this.state.menuState = { menuItems, position };
+  }
+
+  closeMenu() {
+    this.state.menuState = undefined;
+  }
+
+  get groupComponent() {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/components/header_group/header_group_container.ts:68]: it... makes owl crash ??

File: src/plugins/core/header_grouping.ts [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+import { Dimension, HeaderGroup, HeaderIndex, Zone } from "./../../types/misc";
+import { CorePlugin } from "./../core_plugin";
+
+interface State {
+  groups: Record<UID, Record<Dimension, Record<UID, HeaderGroup>>>;
+}
+
+type ToggleHeaderGroupAction = "toggle" | "expand" | "collapse";
+
+export class HeaderGroupingPlugin extends CorePlugin<State> {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/core/header_grouping.ts:18]: Let's not add more code to maintain for something thatwill never be actually useful :upside_down_face: We'll do it if someone is intreseted in the feature

File: src/components/grid_container/grid_container.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  }
+`;
+
+interface Props {
+  gridProps: GridProps;
+  onClick: () => void;
+}
+
+export class GridContainer extends Component<Props, SpreadsheetChildEnv> {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/components/grid_container/grid_container.ts:46]: IMO it makes the 2 layers of `grid-template-columns` (side panel and header groups) easier to understand. Conceptually, it's easier to understand if they are in 2 different components. Forwarding the grid props is wierd tho, I agree.

File: src/plugins/core/header_grouping.ts [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+  private getGroupId(group: HeaderGroup) {
+    return `${group.start}-${group.end}}`;
+  }
+
+  /**
+   * To get layers of groups, and to add/remove headers from groups, we can see each header of a group as a brick. Each
+   * brick falls down in the pile corresponding to its header, until it hits another brick, or the ground.
+   *
+   * With this abstraction, we can very simply group/ungroup headers from groups, and get the layers of groups.
+   * - grouping headers is done by adding a brick to each header pile
```
Comment (LucasLefevre) [inline src/plugins/core/header_grouping.ts:261]: maybe it'd benefit from some ascii art representation for a better understanding. I tried to ask ChatGPT to make it by explaining the context and giving it the algo, but it failed miserably to output something sensible :laughing: 
I can give it a try if you don't want to.

File: src/components/spreadsheet/spreadsheet.xml [hunk 11]
```diff
@@ -14,13 +14,28 @@
...
-          focusComposer="focusGridComposer"
-          exposeFocus="(focus) => this._focusGrid = focus"
-          onComposerContentFocused="() => this.onGridComposerContentFocused()"
-          onGridComposerCellFocused="(content, selection) => this.onGridComposerCellFocused(content, selection)"
-        />
+        <div
+          class="o-grid-container"
+          t-att-class="{'o-two-columns': !sidePanel.isOpen}"
+          t-att-style="gridContainerStyle"
+          t-on-click="() => this.focusGrid()">
```
Comment (LucasLefevre) [inline src/components/spreadsheet/spreadsheet.xml:21]:
```suggestion
          t-on-click="this.focusGrid">
```

File: src/components/header_group/header_group_container.xml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            </t>
+          </t>
+        </div>
+      </div>
+
+      <Menu
+        t-if="state.menuState"
+        menuItems="state.menuState.menuItems"
+        position="state.menuState.position"
+        onClose="() => this.closeMenu()"
```
Comment (LucasLefevre) [inline src/components/header_group/header_group_container.xml:65]:
```suggestion
        onClose.bind="this.closeMenu"
```

File: src/components/header_group/header_group_container.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      margin-left: -2px;
+    }
+  }
+`;
+
+export class HeaderGroupContainer extends Component<Props, SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-HeaderGroupContainer";
+  static components = { RowGroup, ColGroup, Menu };
+
+  state: State = useState({ menuState: undefined });
```
Comment (LucasLefevre) [inline src/components/header_group/header_group_container.ts:47]: why `this.state.menuState` instead of directly `this.menuState` ?

File: src/components/grid/grid.ts [hunk 14]
```diff
@@ -656,6 +676,105 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
...
+    }
+  }
+
+  private processHeaderGroupingEventOnWholeSheet(direction: Direction) {
+    const sheetId = this.env.model.getters.getActiveSheetId();
+    if (direction === "up") {
+      this.env.model.dispatch("COLLAPSE_ALL_HEADER_GROUPS", { sheetId, dimension: "ROW" });
+      this.env.model.dispatch("COLLAPSE_ALL_HEADER_GROUPS", { sheetId, dimension: "COL" });
+    } else if (direction === "down") {
+      this.env.model.dispatch("EXPAND_ALL_HEADER_GROUPS", { sheetId, dimension: "ROW" });
```
Comment (LucasLefevre) [inline src/components/grid/grid.ts:728]: naming detail: maybe name those commands `FOLD_*` and `UNFOLD_*` to be consistent with the `isFolded` property used a bit everywhere?

File: src/components/grid/grid.ts [hunk 15]
```diff
@@ -73,15 +80,23 @@ import { ComposerFocusType } from "../spreadsheet/spreadsheet";
...
 const registries = {
   ROW: rowMenuRegistry,
   COL: colMenuRegistry,
   CELL: cellMenuRegistry,
+  GROUP_HEADERS: groupHeadersMenuRegistry,
+  UNGROUP_HEADERS: unGroupHeadersMenuRegistry,
 };
 
-interface Props {
+export interface GridProps {
```
Comment (LucasLefevre) [inline src/components/grid/grid.ts:99]: this change can be reverted

File: tests/plugins/evaluation_formula_array.test.ts [hunk 16]
```diff
@@ -509,10 +509,10 @@ describe("evaluate formulas that return an array", () => {
       setCellContent(model, "C1", "=MFILL(3,3, 42)");
       expect(getEvaluatedCell(model, "C1").value).toBe("#ERROR");
 
-      addColumns(model, "after", "D", 1);
+      addColumns(model, "after", "C", 1);
```
Comment (LucasLefevre) [inline tests/plugins/evaluation_formula_array.test.ts:512]: what's the reason behind these changes in those tests?

File: src/types/workbook_data.ts [hunk 17]
```diff
@@ -32,6 +32,8 @@ export interface FigureData<T> {
   data: T;
 }
 
+export type HeaderGroupData = Omit<HeaderGroup, "id">;
```
Comment (LucasLefevre) [inline src/types/workbook_data.ts:35]: removed the id :partying_face:  => `HeaderGroupData` can be removed
```suggestion
```

**Review by LucasLefevre — COMMENTED** [review]
another round of comments :)

File: src/types/commands.ts [hunk 18]
```diff
@@ -496,6 +514,47 @@ export interface UpdateLocaleCommand {
...
+
+export interface UnGroupHeadersCommand extends SheetDependentCommand {
+  type: "UNGROUP_HEADERS";
+  dimension: Dimension;
+  start: HeaderIndex;
+  end: HeaderIndex;
+}
+
+export interface ToggleHeaderGroupCommand extends SheetDependentCommand {
+  type: "TOGGLE_HEADER_GROUP";
```
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply src/types/commands.ts:532]: ...or a single command with a boolean param (but I don't know how they would be named)

File: src/plugins/core/header_grouping.ts [hunk 19]
```diff
@@ -0,0 +1,523 @@
...
+        const layers = this.getGroupsLayers(cmd.sheetId, cmd.dimension);
+        if (action === "collapse") {
+          layers.reverse();
+        }
+        const groups = layers.flat();
+        const start = cmd.dimension === "ROW" ? cmd.zone.top : cmd.zone.left;
+        const end = cmd.dimension === "ROW" ? cmd.zone.bottom : cmd.zone.right;
+
+        const groupsToToggle = new Set<HeaderGroup>();
+        for (let header = start; header <= end; header++) {
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/core/header_grouping.ts:140]: Yeah range is slighly more readable, but we've seen that it sucks performance-wise. Here it doesn't really matter, but I'm for avoiding using range, because if we see `range` everywhere we will also use it for methods whose performances matters. 

Just my 2 cents, I can change it if you really want.

File: src/plugins/core/header_grouping.ts [hunk 20]
```diff
@@ -0,0 +1,523 @@
...
+            }
+            groupsToToggle.add(group);
+            break;
+          }
+        }
+
+        for (const group of groupsToToggle) {
+          action === "expand"
+            ? this.unfoldHeaderGroup(cmd.sheetId, cmd.dimension, group)
+            : this.foldHeaderGroup(cmd.sheetId, cmd.dimension, group);
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/core/header_grouping.ts:157]: Really ? I kinda like ternary for one-liner statements, I think it's pretty. I changed it the way you suggested, but I liked it better before TBH.

File: src/components/header_group/header_group_container.ts (hunk 13 above)
Comment (hokolomopo) [reply src/components/header_group/header_group_container.ts:47]: Having `this.menuState` means that I need to have a boolean `isMenuActive` in the `menuState` to toggle the menu, rather than having `this.state.menuState` be undefined.

But I agree that it's ugly. Not sure which I prefer. I can change if you want.

Comment (LucasLefevre) [reply src/components/header_group/header_group_container.ts:47]: I would go with an `isMenuActive`. It also better from semantic POV to have a real boolean rather than comparing with undefined.
Btw, `menuState` could even be named `menu`. `this.menu.isActive`, `this.menu.position`, etc. is looking nice

File: tests/plugins/evaluation_formula_array.test.ts (hunk 16 above)
Comment (hokolomopo) [reply tests/plugins/evaluation_formula_array.test.ts:512]: The allowDispatch of `ADD_COLUMNS_ROWS` was wrong (it used `>`instead of `>=`). I incidentally fixezd it in the PR, and those tests were adding columns out of the sheet (the sheet had 3 columns A,B,C and we tried to add a header after D)

File: src/plugins/core/header_grouping.ts [hunk 21]
```diff
@@ -0,0 +1,523 @@
...
+        if (action === "collapse") {
+          layers.reverse();
+        }
+        const groups = layers.flat();
+        const start = cmd.dimension === "ROW" ? cmd.zone.top : cmd.zone.left;
+        const end = cmd.dimension === "ROW" ? cmd.zone.bottom : cmd.zone.right;
+
+        const groupsToToggle = new Set<HeaderGroup>();
+        for (let header = start; header <= end; header++) {
+          const matchedGroups = groups.filter((g) => g.start - 1 <= header && header <= g.end);
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply src/plugins/core/header_grouping.ts:141]: Your proposal break tests :/ How about a small comment to explain the -1 ?

File: src/plugins/core/header_grouping.ts (hunk 10 above)
Comment (hokolomopo) [reply src/plugins/core/header_grouping.ts:261]: Done 👍

**Comment by hokolomopo** [comment]
@LucasLefevre I rebased, Fixed your comments, added a pretty ASCII art, and created separate commits for various fixes

File: src/plugins/core/header_grouping.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply src/plugins/core/header_grouping.ts]: ok

**Review by LucasLefevre — APPROVED** [review]
robodoo r+ rebase-ff
Thanks for the nice work !
